core/headerchain.go

```diff
@@ -436,15 +439,17 @@ func (hc *HeaderChain) SetGenesis(head *types.Header) {
 //
 // headerValidator implements HeaderValidator.
 type headerValidator struct {
-	hc  *HeaderChain // Canonical header chain
-	Pow pow.PoW      // Proof of work used for validating
+	config *ChainConfig
+	hc     *HeaderChain // Canonical header chain
+	Pow    pow.PoW      // Proof of work used for validating
 }
 
 // NewBlockValidator returns a new block validator which is safe for re-use
-func NewHeaderValidator(chain *HeaderChain, pow pow.PoW) HeaderValidator {
+func NewHeaderValidator(config *ChainConfig, chain *HeaderChain, pow pow.PoW) HeaderValidator {
```

Comment:
As discussed, the config field for this method can be retrieved from the headerchain, so duplicating the data would just open up the potential for strange programming errors when the two mismatch. Please from config and fetch it from headerchain.